ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -393,19 +403,25 @@ func (r *RayJobReconciler) getSubmitterTemplate(rayJobInstance *rayv1.RayJob) (v
 }
 
 // createNewK8sJob creates a new Kubernetes Job. It returns the Job's name and a boolean indicating whether a new Job was created.
-func (r *RayJobReconciler) createNewK8sJob(ctx context.Context, rayJobInstance *rayv1.RayJob, submitterTemplate v1.PodTemplateSpec, rayClusterInstance *rayv1.RayCluster) (string, bool, error) {
+func (r *RayJobReconciler) createNewK8sJob(ctx context.Context, rayJobInstance *rayv1.RayJob, submitterTemplate v1.PodTemplateSpec) (string, bool, error) {
 	job := &batchv1.Job{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      rayJobInstance.Name,
 			Namespace: rayJobInstance.Namespace,
+			Labels: map[string]string{
+				common.KubernetesCreatedByLabelKey: common.ComponentName,
+			},
 		},
 		Spec: batchv1.JobSpec{
-			Template: submitterTemplate,
+			// The `ray job submit` command fails when submission is retried with the same job ID.
+			// Let's disable backoff retry, until job submission retry effectively works.
+			BackoffLimit: pointer.Int32(0),
```

Comment:
We can reduce it, but there will still be cases where the RayCluster doesn't receive the request from the K8s Job.

Reply:
This may depending on the evolution of the retry mechanism longer term, but pragmatically, in the short term, I guess it depends on how likely pre-submission failures are compared to (recoverable) post-submission ones. If the former are more likely, then it'd probably be useful to keep retrying. So maybe a middle ground is to reduce the number of retries from the default, six, to something like two or three. Still it feels like it should be possible to re-submit a job, leveraging the override option from the Ray job manager.

Reply:
> So maybe a middle ground is to reduce the number of retries from the default, six, to something like two or three.

Makes sense.

> Still it feels like it should be possible to re-submit a job, leveraging the override option from the Ray job manager.

I cannot get the point. Would you mind explaining it?

Reply:
>> Still it feels like it should be possible to re-submit a job, leveraging the override option from the Ray job manager.

> I cannot get the point. Would you mind explaining it?

If the submitter job fails "post-submission", i.e., after the first successful `ray job submit` command attempt, any subsequent retry will fail with the `Job with submission_id [...] already exists.` error. If the command could be changed to `ray job submit --overwrite=true`, then it'll fix the issue.

It seems the Ray Job manager has an `overwrite` option: https://github.com/ray-project/ray/blob/56affb7e4b5af8b1da7a322fdc6ecc7e2258d522/dashboard/modules/job/job_manager.py#L919, though it's not currently exposed in the API and CLI.

Reply:
I replied here: https://github.com/ray-project/kuberay/pull/1539#pullrequestreview-1690746497

Reply:
> The default submitter pod template has RestartPolicyNever, do you happen to know why it was still getting retried before this PR?

@architkulkarni `RestartPolicy` controls the kubelet behaviour when a container exits within a Pod. With  `RestartPolicyNever` the Pod is failed, while with `RestartPolicyAlways` for example, the container is restarted in-place. `BackoffLimit` on the hand controls the Job controller behaviour on Pod failure, specifically the maximum number of time a Job Pod should be recreated before failing the Job. So you can have `RestartPolicyNever` and  still have the ray job submission retried when `BackoffLimit` is not zero.